src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   dependent promises.
+
+   Some of these dependencies are explicit to Lwt, e.g. the callbacks registered
+   by [Lwt.bind]. Others are not visible to Lwt, because the user can always
+   register a callback using a function like [Lwt.on_success], and use that
+   callback to complete another initial promise. All the explicit dependencies
+   are created by Lwt's own sequential and concurrent composition functions
+   (so, [Lwt.bind], [Lwt.join], etc). Whether dependencies are explicit or not
+   is relevant only to cancelation.
+
+
+   5. Cancelation
+
+   As described above, ordinary promise completion procedes from an initial
+   promise, forward along callbacks through the dependency graph. Since it
+   starts from an initial promise, it can only be triggered using a resolver.
+
+   Cancelation is a sort of dual to ordinary completion. Instead of starting at
+   an initial promise/resolver, cancelation starts at *any* promise. It then
+   goes *backwards* through the explicit dependency graph, looking for
+   cancelable initial promises to cancel -- those that were created by
+   [Lwt.task]. After finding them, cancelation completes them normally with
+   [Failed Lwt.Canceled], causing an ordinary promise completion process.
+
+   To summarize, cancelation is a way to trigger an *ordinary* completion of
+   promises created with [Lwt.task], by first searching for them in the promise
+   dependency graph (which is assembled by [Lwt.bind], [Lwt.join], etc).
+
+   This backwards search is triggered only by [Lwt.cancel]. It is also possible
+   for the user to cancel a promise directly by failing it with [Lwt.Canceled],
+   but in all cases where the user can do so, the search would be redundant
+   anyway -- the user has only two ways of directly failing a promise with
+   [Lwt.Canceled] (or any exception, for that matter):
+
+   - The user can create an initial promise, then fail it through its resolver.
+     The search is redundant because it would find only the same initial promise
+     to cancel.
+   - The user can create a trivial promise by calling [Lwt.fail Lwt.Canceled].
+     The search is again redundant; in this case it would find nothing to
+     cancel.
+
+   Note that there is a quirk: only promises created by [Lwt.task] are
+   susceptible to being canceled by [Lwt.cancel], but the user can manually
+   cancel initial promises created by both [Lwt.task] and [Lwt.wait].
+
+   Due to [Lwt.cancel], promise cancelation, and therefore completion, can be
+   initiated by the user without access to a resolver. This is important for
+   reasoning about state changes in the implementation of Lwt, and is referenced
+   in some implementation detail comments.
+
+
+   6. No I/O
+
+   The Lwt core deliberately doesn't do I/O. The completion loop stops running
+   once no promises can be completed immediately. It has to be restarted later
+   by some external process.
```

Comment:
IIUC, this is for the scheduler/IO-loop to manage. Typically, when you do something like `Lwt_unix.read ()` it
- creates a promise (`p`),
- then it prepares a callback (`c`) that, when triggered, completes the promise,
- then it calls `select`, and it sets `c` to be called when `select` is triggered when uses `select` returns.

Note: it can be select or it can be some other similar mechanism.  
Note: the upcoming algebraic effects might make it easier to integrate Lwt to an IO-loop.  
Note: the callback `c` is not a callback set by `Lwt`, it is set by `Lwt_unix` (in this example, other libs can do something similar)

This is from what I remember from discussions about Lwt several years ago. It might be inaccurate, in which case, I'd be interested in what actually happens.

I think that there should be a reference like “check the documentation of the file Lwt_unix.ml for an explanation of one IO-loop” and then a detailed explanation in that file.